src/text.rs

```diff
@@ -455,3 +465,68 @@ where
         self.lines.extend(lines);
     }
 }
+
+impl fmt::Write for Text<'_> {
+    #[inline]
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        /// We would like the compiler to inline the return line,
+        /// and therefore see that the result is always `Ok(())`,
+        /// allowing it to turn `unwrap()` calls into a no-op.
+        /// However, we don't want to inline
+        /// the entire `write_str` function. Hence this split.
+        #[inline(never)]
+        fn write_str_inner(text: &mut Text<'_>, s: &str) {
+            // Split on newlines.
+            let mut lines = s.split('\n').map(|s| s.strip_suffix('\r').unwrap_or(s));
+
+            // Append the start of the to-be-written text
+            // (before the first newline) to the end of the previous line,
+            // if there is one.
+            if let Some(prev_end) = text.lines.last_mut() {
+                if let Some(first) = lines.next() {
+                    if !first.is_empty() {
+                        // Only make a new `String` if the trailing one is not suitable to push onto.
+                        match prev_end.spans.last_mut() {
```

Comment:
I'm probably being a bit pedantic here, but I still don't quite fully understand this just from reading this code. The comments describe how this does what it's doing, but they don't describe what the intent of doing that is.

I think it's probably something like, write to the text, maintaining any new lines when they are written, but append text to the end of any non-terminated newline. That's a pretty bad explanation, and I'm not entirely certain it is correct. Please help by filling that out. Think about what you'd want to see in the docs.rs page for this method.

Without telling me what this is supposed to do, it's difficult to review it for whether it does the thing that it's saying it will do. How should it act, what are the edge cases, etc. Even the comments seem to imply some context that I'm just not aware of reading this code.

This sounds minor, but it's important to be able to maintain the code going forward, and capturing intent is key in doing that. The how part is right there in the code. if all the comments were removed we'd still be able to work that out, but the intent isn't there. We'd have to guess or search commits and PRs for that info.